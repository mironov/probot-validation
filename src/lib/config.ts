import { Group, Rule } from "./rules";

export interface Config {
  groups: Group[];
}

const mergeLabelOnError = (...inputs: any[]): string[] => {
  const labels = new Set<string>();
  inputs.forEach(input => {
    if (input instanceof Array) {
      input.forEach(label => labels.add(label));
    } else if (typeof input === "string") {
      labels.add(input);
    } else if (input != null) {
      throw new Error(`Unexpected value for 'labelOnError': ${input}`);
    }
  });
  return [...labels];
};

export const parseConfig = (configInput: any): Config => {
  const config: Config = { groups: [] };
  if (typeof configInput !== "object") {
    throw new Error("Top-level configuration must be an object");
  }

  if (!(configInput.groups instanceof Array)) {
    throw new Error("Top-level 'groups' config must be an array");
  }

  configInput.groups.forEach((groupConfig: any) => {
    if (typeof groupConfig != "object") {
      throw new Error(
        `Expected group configuration to be an object, got: ${JSON.stringify(
          groupConfig
        )}`
      );
    }

    let { context, filter, rules } = groupConfig;

    if (typeof context != "string") {
      throw new Error("'context' must be present and a string");
    }

    if (!(rules instanceof Array)) {
      throw new Error("'rules' must be present and an array");
    }

    rules = rules.map(
      (ruleInput): Rule => {
        if (ruleInput.labelOnError instanceof Array) {
        }

        return {
          context,
          labelOnError: mergeLabelOnError(
            groupConfig.labelOnError,
            ruleInput.labelOnError
          ),
          test: ruleInput.test as string,
          message: ruleInput.message as string,
          level: ruleInput.level || "error"
        };
      }
    );

    config.groups.push({
      filter,
      rules
    });
  });

  return config;
};
