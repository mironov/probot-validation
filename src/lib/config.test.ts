import "jest";
import { parseConfig } from "./config";
import yaml from "js-yaml";
import fs from "fs";
import path from "path";

const loadFixtureYAML = (fixturePath: string) =>
  yaml.safeLoad(
    fs
      .readFileSync(path.resolve(__dirname, "./__fixtures__", fixturePath))
      .toString()
  );

describe("parseConfig", () => {
  test("when given an invalid top-level type", () => {
    expect(() => {
      parseConfig("foo");
    }).toThrowError("Top-level configuration must be an object");
  });

  test("when missing groups", () => {
    expect(() => {
      parseConfig({});
    }).toThrowError("Top-level 'groups' config must be an array");
  });

  test("when given an invalid group", () => {
    expect(() => {
      parseConfig({ groups: ["foo", "bar"] });
    }).toThrowError('Expected group configuration to be an object, got: "foo"');
  });

  test("when given a valid configuration", () => {
    expect(parseConfig(loadFixtureYAML("valid-config.yaml"))).toMatchSnapshot();
  });
});
