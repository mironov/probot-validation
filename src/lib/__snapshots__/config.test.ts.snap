// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseConfig when given a valid configuration 1`] = `
Object {
  "groups": Array [
    Object {
      "filter": "$not(pull_request)",
      "rules": Array [
        Object {
          "context": "context1",
          "labelOnError": Array [
            "label1",
          ],
          "level": "error",
          "message": "foo error 1",
          "test": "foo.x and foo.y",
        },
        Object {
          "context": "context1",
          "labelOnError": Array [
            "label1",
          ],
          "level": "error",
          "message": "bar error 1",
          "test": "bar.x or bar.y",
        },
      ],
    },
    Object {
      "filter": undefined,
      "rules": Array [
        Object {
          "context": "context2",
          "labelOnError": Array [
            "work-in-progress",
            "blocked",
          ],
          "level": "warn",
          "message": "Pull request is a work in progress",
          "test": "$match(title, /.*WIP.*/)",
        },
      ],
    },
    Object {
      "filter": undefined,
      "rules": Array [
        Object {
          "context": "context3",
          "labelOnError": Array [
            "label1",
            "label2",
            "label3",
          ],
          "level": "error",
          "message": "error",
          "test": "bar.x",
        },
      ],
    },
  ],
}
`;
